File: Firefly Original/eclipse workspace/Firefly-common/src/animation/primitives/ConcentricSpheresPrim.cpp
/*
 * ConcentricSpheresPrim.cpp
 *
 *  Created on: Aug 6, 2016
 *      Author: Denis
 */

#include <animation/primitives/ConcentricSpheresPrim.h>

glm::vec4 ConcentricSpheresPrim::renderAtDistanceFromPoint(double distance, ArbitraryMap* details) {
	double distanceIntoSphere = sizeInRings * ringRadius - distance;

	slicer.interval = ringRadius;
	int ring = slicer.getInterval(distanceIntoSphere);
	double phase = slicer.getPhase(distanceIntoSphere);

	int colorIndex = ring % colors.size();
	int blendColorIndex = phase <= 0.5 ? (ring-1) % colors.size() : (ring+1) % colors.size();

	float colorBrightness = fmax(0.0, slicer.getSymmetricalPhase(distanceIntoSphere) + colorBlendingFactor) / (1.0 + colorBlendingFactor);
	float blendColorBrightness = fmax(0.0,colorBlendingFactor - slicer.getSymmetricalPhase(distanceIntoSphere)) / (1.0 + colorBlendingFactor);

	glm::vec4 color;
	if(ring < 0 || ring > finalRing) {
		color = glm::vec4(0.0, 0.0, 0.0, 0.0);
	} else if ((ring == 0 && phase < 0.5) || (ring == finalRing && phase > 0.5)) {
		color = colors[colorIndex];
		color.a = fmin(colorBrightness, (ring == 0 ? phase : 1.0-phase)/0.5);
	} else {
		color = colors[colorIndex]*(1.0f-blendColorBrightness) + colors[blendColorIndex]*blendColorBrightness;
		color *= colorBrightness + blendColorBrightness;
		color += glm::vec4(0.0, 0.0, 0.0, CSP_MIN_ALPHA) * (1.0f - colorBrightness - blendColorBrightness);
	}

	float distToEdge = fmin((float)ring + phase, (float)finalRing+1.0-(float)ring-phase);
	float edgeAlphaMultiplier = fmax(0.0, fmin(1.0, distToEdge/CSP_EDGE_FADE_SIZE));
	color.a *= edgeAlphaMultiplier;

	if(colorBrightness > 0.0 && ring >= 0) {
		details->setInt("shaderIndex", ring);
	}
	return color;
}

ConcentricSpheresPrim::~ConcentricSpheresPrim() {
	// TODO Auto-generated destructor stub
}

File: Firefly/Firefly Controller/Assets/Scripts/Animation/Primitives/ConcentricSpheresPrim.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Firefly
{
    public class ConcentricSpheresPrim : PointDistancePrimitive
    {
        public const double CSP_MIN_ALPHA = 0.85;
        public const double CSP_EDGE_FADE_SIZE = 3.0;

        public List<Vector4> colors;
        public double colorBlendingFactor;
        public double ringRadius, sizeInRings;
        public int finalRing = int.MaxValue;

        protected IntervalSlicerWithSymmetricalPhases slicer;

        public ConcentricSpheresPrim(
            Vector3 centerPoint,
            double ringRadius,
            double sizeInRings,
            List<Vector4> colors,
            double colorBlendingFactor)
            : base(centerPoint)
        {
            this.ringRadius = ringRadius;
            this.sizeInRings = sizeInRings;
            this.colors = colors;
            this.colorBlendingFactor = colorBlendingFactor;
            this.slicer = new IntervalSlicerWithSymmetricalPhases(0.0, ringRadius);
        }

        protected override Vector4 RenderAtDistanceFromPoint(double distance, ArbitraryMap details)
        {
            double distanceIntoSphere = sizeInRings * ringRadius - distance;

            slicer.interval = ringRadius;
            int ring = slicer.GetInterval(distanceIntoSphere);
            double phase = slicer.GetPhase(distanceIntoSphere);

            // FireflyUtils.Mod rather than % — the C++ indexed with a possibly-negative
            // result, which C# would throw on. See Port Notes.
            int colorIndex = FireflyUtils.Mod(ring, colors.Count);
            int blendColorIndex = phase <= 0.5
                ? FireflyUtils.Mod(ring - 1, colors.Count)
                : FireflyUtils.Mod(ring + 1, colors.Count);

            float colorBrightness = (float)(Math.Max(0.0, slicer.GetSymmetricalPhase(distanceIntoSphere) + colorBlendingFactor) / (1.0 + colorBlendingFactor));
            float blendColorBrightness = (float)(Math.Max(0.0, colorBlendingFactor - slicer.GetSymmetricalPhase(distanceIntoSphere)) / (1.0 + colorBlendingFactor));

            Vector4 color;
            if (ring < 0 || ring > finalRing)
            {
                color = new Vector4(0.0f, 0.0f, 0.0f, 0.0f);
            }
            else if ((ring == 0 && phase < 0.5) || (ring == finalRing && phase > 0.5))
            {
                color = colors[colorIndex];
                color.w = (float)Math.Min(colorBrightness, (ring == 0 ? phase : 1.0 - phase) / 0.5);
            }
            else
            {
                color = colors[colorIndex] * (1.0f - blendColorBrightness) + colors[blendColorIndex] * blendColorBrightness;
                color *= colorBrightness + blendColorBrightness;
                color += new Vector4(0.0f, 0.0f, 0.0f, (float)CSP_MIN_ALPHA) * (1.0f - colorBrightness - blendColorBrightness);
            }

            float distToEdge = (float)Math.Min((double)ring + phase, (double)finalRing + 1.0 - (double)ring - phase);
            float edgeAlphaMultiplier = (float)Math.Max(0.0, Math.Min(1.0, distToEdge / CSP_EDGE_FADE_SIZE));
            color.w *= edgeAlphaMultiplier;

            if (colorBrightness > 0.0 && ring >= 0)
            {
                details.SetInt("shaderIndex", ring);
            }
            return color;
        }
    }
}
